File: src/python/problems/AddStrings.py
# https://leetcode.com/problems/add-strings/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class AddStrings:
    def Solution(self, num1: str, num2: str) -> list[list[int]]:
        length = max(len(num1), len(num2))

        diff = '0' * abs(len(num1) - len(num2))
        if len(num1) < len(num2):
            num1 = diff + num1
        else:
            num2 = diff + num2

        result = ""
        carry = 0
        for i in range(length - 1, -1, -1):
            sum = carry + int(num1[i]) + int(num2[i])
            carry = sum // 10
            result = str(sum % 10) + result

        return str(carry) + result if carry > 0 else result



AddStrings().Solution("154", "12")

File: src/csharp/Problems/AddStrings.cs
//https://leetcode.com/problems/add-strings/

namespace LeetCode.Problems;

public sealed class AddStrings : ProblemBase
{
    [Theory]
    [ClassData(typeof(AddStrings))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.Param("10").Param("12").Result("22"))
          .Add(it => it.Param("152").Param("12").Result("164"))
          .Add(it => it.Param("15").Param("122").Result("137"))
          .Add(it => it.Param("15").Param("0").Result("15"))
          .Add(it => it.Param("0").Param("15").Result("15"))
          .Add(it => it.Param("1233").Param("15").Result("1248"))
          .Add(it => it.Param("15").Param("12").Result("27"));

    private string Solution(string num1, string num2)
    {
        var length = Math.Max(num1.Length, num2.Length);

        var diff = new string(Enumerable.Repeat('0', Math.Abs(num2.Length - num1.Length)).ToArray());
        if (num1.Length < num2.Length)
        {
            num1 = diff + num1;
        }
        else
        {
            num2 = diff + num2;
        }

        var result = new StringBuilder();
        var carry = 0;
        for (var i = length - 1; i >= 0; i--)
        {
            var add = carry + (num1[i] - '0') + (num2[i] - '0');
            carry = add / 10;
            add %= 10;

            result.Insert(0, add);
        }

        if (carry > 0)
        {
            result.Insert(0, carry);
        }

        return result.ToString();
    }
}